js/i18n.js

```diff
@@ -8,19 +8,33 @@
     var messages = window.config.localeMessages;
     var locale = window.config.locale;
 
-    window.i18n = function (message, substitutions) {
-      if (!messages[message]) {
+    window.i18n = function (key, values) {
+      var data = messages[key];
+      if (!data) {
         return;
       }
-      var s = messages[message].message;
-      if (substitutions instanceof Array) {
-        substitutions.forEach(function(sub) {
-          s = s.replace(/\$.+?\$/, sub);
-        });
-      } else if (substitutions) {
-        s = s.replace(/\$.+?\$/, substitutions);
+
+      var message = data.message;
+      var placeholders = data.placeholders;
+      if (!values || !placeholders) {
+        return message;
+      }
+
+      if (!(values instanceof Array)) {
+        values = [values];
+      }
+
+      var names = Object.keys(placeholders);
```

Comment:
This is potentially brittle because `Object.keys` doesn't guarantee any particular order for arbitrary keys. In fact, it will attempt to return the keys in sorted order if you have any keys which are non-negative integers.

I think we can get away with it given our current sent of strings, but chrome's i18n, for example, solves this by using "$N" in the `content` field to map a certain placeholder to the Nth incoming argument:

```
"placeholders": {
  "name": {
    "content": "$1",
    "example": "Bob"
  }
}
```